To: vim_dev@googlegroups.com
Subject: Patch 7.3.404
Fcc: outbox
From: Bram Moolenaar <Bram@moolenaar.net>
Mime-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
------------

Patch 7.3.404
Problem:    When a complete function uses refresh "always" redo will not work
	    properly.
Solution:   Do not reset compl_leader when compl_opt_refresh_always is set.
	    (Yasuhiro Matsumoto)
Files:	    src/edit.c


*** ../vim-7.3.403/src/edit.c	2012-01-10 22:26:12.000000000 +0100
--- src/edit.c	2012-01-20 14:26:06.000000000 +0100
***************
*** 3465,3475 ****
      if (ins_compl_need_restart())
  	ins_compl_restart();
  
!     vim_free(compl_leader);
!     compl_leader = vim_strnsave(ml_get_curline() + compl_col,
  				     (int)(curwin->w_cursor.col - compl_col));
!     if (compl_leader != NULL)
! 	ins_compl_new_leader();
  }
  
  /*
--- 3465,3481 ----
      if (ins_compl_need_restart())
  	ins_compl_restart();
  
!     /* When 'always' is set, don't reset compl_leader. While completing,
!      * cursor don't point original position, changing compl_leader would
!      * break redo. */
!     if (!compl_opt_refresh_always)
!     {
! 	vim_free(compl_leader);
! 	compl_leader = vim_strnsave(ml_get_curline() + compl_col,
  				     (int)(curwin->w_cursor.col - compl_col));
! 	if (compl_leader != NULL)
! 	    ins_compl_new_leader();
!     }
  }
  
  /*
***************
*** 4554,4559 ****
--- 4560,4570 ----
      int	    found_end = FALSE;
      int	    advance;
  
+     /* When user complete function return -1 for findstart which is next
+      * time of 'always', compl_shown_match become NULL. */
+     if (compl_shown_match == NULL)
+ 	return -1;
+ 
      if (compl_leader != NULL
  			&& (compl_shown_match->cp_flags & ORIGINAL_TEXT) == 0)
      {
*** ../vim-7.3.403/src/version.c	2012-01-20 13:39:03.000000000 +0100
--- src/version.c	2012-01-20 14:31:26.000000000 +0100
***************
*** 716,717 ****
--- 716,719 ----
  {   /* Add new patch number below this line */
+ /**/
+     404,
  /**/

-- 
ARTHUR: I've said I'm sorry about the old woman, but from the behind you
        looked ...
DENNIS: What I object to is that you automatically treat me like an inferior...
ARTHUR: Well ... I AM king.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
